<a href="https://colab.research.google.com/github/swaroopa1352/NNSwaroopa/blob/main/M1_Assignment_Stock_Price_Prediction_with_LLM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module 1 No Code/Low Code Assignment: Stock Price Prediction with LLM (Using Gemini in Colab)

This assignment explores how large language models (LLMs) can perform complex tasks that previously required coding. Instead of manually writing code, we will interact with Google's Gemini (Click on the Gemini icon on the top right corner) in Google Colab to:

- Load and analyze stock price data.
- Engineer relevant features for prediction.
- Build and evaluate models such as regression and random forest.
- Discuss potential trading strategies based on predictions.

**Note**: While this notebook contains example code, students are encouraged to use LLMs to generate code snippets and explanations, managing interactions within the LLM's context window.

## Task Breakdown

### Step 1: Data Exploration
- **Prompt the LLM**: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance.
- **Key Considerations**: Ensure the model retains context while summarizing key statistics.

### Step 2: Feature Engineering
- **Prompt the LLM**: Generate meaningful features such as moving averages, technical indicators, and volatility measures.
- **Managing Context**: If the LLM forgets prior outputs, remind it by reintroducing key statistics.

### Step 3: Model Building
- **Prompt the LLM**: Train a model (regression for price prediction, random forest for trend classification).
- **Split Ratio Consideration**: Explain the importance of train-test split (e.g., 70/30) and verify the LLM respects it.

### Step 4: Model Evaluation & Prediction
- **Prompt the LLM**: Use the trained model to predict stock prices for 2023-2024.
- **Analyze Performance**: Interpret accuracy, R² scores, and classification reports.

### Step 5: Strategy Discussion
- **Prompt the LLM**: Based on model performance, discuss possible trading strategies.
- **Critical Thinking**: Evaluate whether the model is truly useful for financial decision-making.

---

Throughout the assignment, keep in mind that LLMs have a limited context window. If necessary, break down tasks into smaller chunks and reintroduce critical information.



## Using Gemini in Colab

To use Gemini directly in your browser within Google Colab:

1. **Create a new code cell** by clicking the `+ Code` button.
2. In the newly created cell, you will see a placeholder text: **"Start coding or generate with AI"**.
3. Click on **"Generate"** and enter your prompt, such as:
   > "Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics."
4. Gemini will generate code based on your prompt. You can then review, modify, and run it within the notebook.
5. If the model loses context, you may need to remind it of previous responses or break down tasks into smaller steps.

This method allows you to interactively generate and refine code without manually writing every line.


In [6]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Display summary statistics
print(data.describe())


/tmp/ipython-input-6-3008817219.py:9: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Shape after loading data: (756, 5)
Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.633102  129.181519  125.971433  127.555151  1.120920e+08
std      30.421462   30.698898   30.139549   30.449562  5.602586e+07
min      54.378582   55.379535   51.528416   55.277744  3.519590e+07
25%     112.709524  114.204935  111.251838  112.889681  7.636470e+07
50%     133.010757  134.143870  131.050650  132.832284  9.493580e+07
75%     148.221519  149.388588  146.276903  147.588770  1.296327e+08
max     178.645645  179.558457  175.809061  179.254190  4.265100e+08


In [7]:
# Prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01
# using Yahoo Finance and summarize its key statistics.

import yfinance as yf

data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())



/tmp/ipython-input-7-3699561310.py:6: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.633102  129.181519  125.971433  127.555151  1.120920e+08
std      30.421462   30.698898   30.139549   30.449562  5.602586e+07
min      54.378582   55.379535   51.528416   55.277744  3.519590e+07
25%     112.709524  114.204935  111.251838  112.889681  7.636470e+07
50%     133.010757  134.143870  131.050650  132.832284  9.493580e+07
75%     148.221519  149.388588  146.276903  147.588770  1.296327e+08
max     178.645645  179.558457  175.809061  179.254190  4.265100e+08


## Now try doing step 2 - step 5 by yourself

The sample prompts and LLM responses are listed below for your reference.

Here are some additional features that might help improve a stock price prediction model:

1.  **More Moving Averages:** Experiment with different window sizes (e.g., 20-day, 100-day, 200-day moving averages).
2.  **Other Technical Indicators:**
    *   Moving Average Convergence Divergence (MACD)
    *   Bollinger Bands
    *   Stochastic Oscillator
    *   Average True Range (ATR)
    *   Ichimoku Cloud
3.  **Lagged Prices:** Include the closing price from previous days as features.
4.  **Lagged Volume:** Include the trading volume from previous days.
5.  **Volatility Measures:**
    *   Parkinson Volatility
    *   Garman-Klass Volatility
6.  **Time-Based Features:**
    *   Day of the week
    *   Month of the year
    *   Quarter of the year
7.  **External Data:**
    *   News sentiment scores related to the stock or the overall market.
    *   Economic indicators (e.g., interest rates, inflation, GDP growth).
    *   Industry or sector specific indices.
    *   Related asset prices (e.g., commodities, currencies).
8.  **Fundamental Data:**
    *   Earnings per share (EPS)
    *   Price-to-earnings (P/E) ratio
    *   Dividend yield
    *   Revenue growth

In [32]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

# Define features (X) and target (y)
# Using the generated features including lagged prices
features = ['Open', 'High', 'Low', 'Close', 'Volume', 'MA_5', 'MA_10', 'MA_50', 'RSI', 'Vol_5', 'Close_Lag1', 'Close_Lag2', 'Close_Lag3']
target = 'Close' # Predict tomorrow's closing price

# Create the target variable (tomorrow's closing price)
data['Target'] = data['Close'].shift(-1)

print("\nShape before dropping NaNs:", data.shape)
print("\nNon-null counts before dropping NaNs:")
print(data.notnull().sum())


# Drop NaN values created by feature engineering, lagged prices and target creation
data_cleaned = data.dropna()

print("\nShape after dropping NaNs:", data_cleaned.shape)


X = data_cleaned[features]
y = data_cleaned['Target']

print("Shape of X:", X.shape)
print("Shape of y:", y.shape)

# Split data into training and testing sets (70/30 ratio)
if data_cleaned.shape[0] > 1:
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

    # Initialize and train the Linear Regression model
    model = LinearRegression()
    model.fit(X_train, y_train)

    # Make predictions on the test set
    y_pred = model.predict(X_test)

    # Evaluate the model
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)

    print(f"Mean Squared Error (MSE): {mse:.2f}")
    print(f"R-squared (R²) Score: {r2:.2f}")
else:
    print("Not enough data after dropping NaNs to perform train-test split and train the model.")


Shape before dropping NaNs: (756, 14)

Non-null counts before dropping NaNs:
Price       Ticker
Close       AAPL      756
High        AAPL      756
Low         AAPL      756
Open        AAPL      756
Volume      AAPL      756
MA_5                  752
MA_10                 747
MA_50                 707
RSI                   755
Vol_5                 752
Close_Lag1            755
Close_Lag2            754
Close_Lag3            753
Target                  0
dtype: int64

Shape after dropping NaNs: (0, 14)
Shape of X: (0, 13)
Shape of y: (0,)
Not enough data after dropping NaNs to perform train-test split and train the model.


In [25]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import pandas as pd

# Create a binary target variable
# 1 if the close price increases tomorrow, 0 otherwise
data['Price_Increase'] = (data['Close'].shift(-1) > data['Close']).astype(int)

print("Shape before dropping NaNs:", data.shape)

# Drop NaN values created by feature engineering and target creation
data_cleaned = data.dropna()

print("Shape after dropping NaNs:", data_cleaned.shape)


# Define features (X) and target (y)
features = ['Open', 'High', 'Low', 'Close', 'Volume', 'MA_5', 'MA_10', 'MA_50', 'RSI', 'Vol_5'] # Using the features engineered earlier
target = 'Price_Increase'

# Check if there are enough samples after dropping NaNs
if data_cleaned.shape[0] > 1:
    X = data_cleaned[features]
    y = data_cleaned[target]

    print("Shape of X:", X.shape)
    print("Shape of y:", y.shape)


    # Split data into training and testing sets (70/30 ratio)
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

    # Initialize and train the Random Forest Classifier model
    model = RandomForestClassifier(n_estimators=100, random_state=42)
    model.fit(X_train, y_train)

    # Make predictions on the test set
    y_pred = model.predict(X_test)

    # Evaluate the model
    accuracy = accuracy_score(y_test, y_pred)
    conf_matrix = confusion_matrix(y_test, y_pred)
    class_report = classification_report(y_test, y_pred)

    print(f"Accuracy: {accuracy:.2f}")
    print("Confusion Matrix:")
    print(conf_matrix)
    print("Classification Report:")
    print(class_report)

    # Report Feature Importance
    feature_importances = pd.Series(model.feature_importances_, index=features).sort_values(ascending=False)
    print("\nFeature Importance:")
    print(feature_importances)
else:
    print("Not enough data after dropping NaNs to perform train-test split and train the model.")

Shape before dropping NaNs: (0, 15)
Shape after dropping NaNs: (0, 15)
Not enough data after dropping NaNs to perform train-test split and train the model.


Accuracy: 0.55
Confusion Matrix:
[[53 49]
 [47 63]]
Classification Report:
              precision    recall  f1-score   support

           0       0.53      0.52      0.52       102
           1       0.56      0.57      0.57       110

    accuracy                           0.55       212
   macro avg       0.55      0.55      0.55       212
weighted avg       0.55      0.55      0.55       212


Feature Importance:
RSI       0.115121
Vol_5     0.109916
Volume    0.109676
MA_50     0.102513
Open      0.097483
High      0.095722
Close     0.095242
MA_10     0.092781
MA_5      0.092103
Low       0.089443
dtype: float64


# Sample Prompts and LLM Responses
---

## Step 1: Data Exploration

### Prompt:
> Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 from Yahoo Finance, keep it in memory and summarize its key statistics.

### LLM Response:
- **Highest Closing Price:** \$182.01 on January 3, 2022
- **Lowest Closing Price:** \$56.01 on March 23, 2020
- **Average Closing Price:** \$126.69
- **Total Trading Days:** 756

---

## Step 2: Feature Engineering

### Prompt:
> Suggest some features that might help improve a stock price prediction model.

### LLM Response:
1. **Historical Prices**: Include past closing, opening, high, and low prices.
2. **Volume**: Daily trading volumes.
3. **Moving Averages**: Simple and exponential moving averages (e.g., 5-day, 10-day, 50-day).
4. **Technical Indicators**: RSI, MACD, Bollinger Bands, etc.
5. **Volatility**: Historical volatility, Average True Range (ATR).
6. **Sentiment Analysis**: News articles, social media sentiment.
7. **Economic Indicators**: Interest rates, inflation rates, GDP growth.
8. **Time-based Features**: Day of the week, month, year.
9. **Sector Performance**: Comparison with industry and sector indices.
10. **External Events**: Major geopolitical events, earnings reports.

---

## Step 3.1: Model Building - Linear Regression

### Prompt:
> Adopt the first four kinds of features and train a regression model to predict tomorrow's stock price. Split your training and testing data by a 70/30 ratio and report your performance metric.

### LLM Response:
- **Mean Squared Error (MSE):** 859.04
- **R-squared (R²) Score:** -0.009

These metrics indicate that the model is not performing well, as an R² score close to zero or negative suggests that the model does not explain the variability in the target variable better than a simple mean of the target variable. This outcome suggests that the chosen features and model might need adjustments or improvements to enhance prediction accuracy. ​​

---
(It might be too hard to predict the exact price but how about ups and downs?)
# Step 3.2: Model Building - Random Forest Classification

### Prompt:
> Create a binary target variable based on historical prices (1 if the close price increase tomorrow, 0 other), and train a Random Forest model using the same train/test split ratio. Report feature importance and performance metrics.

### LLM Response:
- **Accuracy:** 71.82%
- **Confusion Matrix:** [[84, 26],
 [36, 74]]
- **Classification Report:**

              precision    recall  f1-score   support

         0       0.70      0.76      0.73       110
         1       0.74      0.67      0.70       110
        accuracy                     0.72       220
        macro avg    0.72      0.72      0.72       220
        weighted avg  0.72      0.72      0.72       220



- **Feature Importance:** The most important feature for predicting whether the price will increase or decrease the next day is the closing price, followed by the 5-day moving average (MA_5). The model achieved an accuracy of approximately 71.82%, indicating a reasonable performance for this classification task.
         Feature	Importance
         Open	   0.0900
         High	   0.0883
         Low	   0.0980
         Close	   0.3087
         Volume	    0.0982
         MA_5	   0.1242
         MA_10	   0.0974
         MA_50	   0.0951

---

## Step 4: Model Evaluation & Prediction
### Prompt:
> Use this model to predict stock price increase/decrease for period starting from 2023-01-01 to 2024-01-01. Report your performance metrics.

### LLM Response:
The Random Forest classifier's performance on the new test data from 2023-01-01 to 2024-01-01 is as follows:
- **Accuracy:** 60.85%
- **Confusion Matrix:** [[105,   0],
 [ 83,  24]]
- **Classification Report:**

            precision    recall  f1-score   support

       0       0.56      1.00      0.72       105
       1       1.00      0.22      0.37       107

        accuracy                   0.61       212
        macro avg  0.78     0.61       0.54       212
        weighted avg 0.78    0.61        0.54       212

---
## Step 5: Strategy Discussion
### Prompt:
> What do you think about the performance of the model and what trading strategies could I build with this model?

### LLM Response:
- The model shows a moderate performance with an accuracy of approximately 60.85%. However, the precision and recall metrics indicate that while the model is good at identifying days when the stock price does not increase (class 0), it struggles significantly to identify days when the stock price does increase (class 1). This suggests a need for further feature engineering or model tuning to improve the prediction performance for this specific classification task.
- Potential Trading Strategies
Given the model's performance, potential trading strategies could be formulated as follows:

#### Conservative Strategy:

- **Focus**: Only trade on days when the model predicts a price increase with high confidence.
- **Action**: Buy stocks when the model predicts an increase (class 1).
- **Risk Management**: Since the model has a high precision for the no-increase class (class 0), it might be more reliable to stay out of the market when the model predicts no increase.

#### Contrarian Strategy:

- **Focus**: Take advantage of the model's high recall for no price increase predictions (class 0).
- **Action**: Sell short or avoid buying stocks on days when the model predicts no increase.
- **Risk Management**: This strategy assumes that the model’s predictions of no price increase are more reliable. It requires a careful assessment of market conditions to avoid false signals.

#### Hybrid Strategy:

- **Combine**: Use both conservative and contrarian strategies based on confidence thresholds.
- **Action**: Set a higher threshold for buying (only buy when the predicted probability of price increase is significantly high) and a lower threshold for selling or staying out of the market.
- **Risk Management**: Adjust thresholds dynamically based on recent model performance and market volatility.


---
## Conclusion

Through this assignment, we have explored how LLMs can assist in complex machine learning tasks without requiring extensive coding knowledge. Key takeaways:

- **Context Management**: LLMs have limited memory, so breaking tasks into smaller steps and reinforcing context is crucial.
- **Feature Engineering with LLMs**: The model can suggest features but needs guidance on financial relevance.
- **Modeling with LLMs**: LLMs can write and refine code, but human judgment is necessary to validate results.
- **Real-world Limitations**: Stock price prediction remains challenging; models should be used critically and not solely relied upon for trading decisions.

Next Steps:
- Experiment with different LLMs and compare outputs.
- Try refining feature engineering techniques.
- Test alternative modeling approaches, such as deep learning or ensemble methods.

